### ISSUE TITLE
Are the TF derivatives correctly implemented?

### ISSUE THREAD

**Issue by TeHMoroS** [davidrmiller/neural2d#40]
Hello David.

I'm having doubts about the implementation of transfer functions in Neural2D, specifically - their derivatives. So far I've only analyzed and used the TanH and Logistic TFs, but let's get to it.

First of, when feeding forward you call the TF and store it's output for x in the neuron output, so far correct. What I'm having issues with are the derivatives. While they're mathematically correct, I think the code logic might be flawed here. When doing backward propagation gradient calculation (I'm referring to regular layers), you take the said neuron output (which is at that time "tanh(x)" from feeding forward) and use it as a parameter for the TFs derivative, which is also a full math equivalent (so you're calculating: "1.0f - tanh(tanh(x)) * tanh(tanh(x))" - pretty much the same thing happens for the Logistic TF).

That's what I have trouble understanding. Is it a bug that the gradient is double-functioned ("tanh(tanh(x))") or should it be like that on purpose? Is so, that why? Shouldn't the derivatives include the fact, that the basic function has already been called (making tanh's code derivative like: "1.0f - x * x")?

EDIT: I did an experiment for the XOR function. The original code needs over 8000 iterations to solve the problem. After reimplementing the TanH TF derivative as mentioned above, it completed the calculations in 805 iterations only and got a lot more "Correct" outputs than in the original code (I know they're used in classification, but I'm just mentioning it).

**Comment by nobleliar9296** [comment]
hello sir,
            i am not David but derivative tells you about the rate of
change of a fn'. Do see up linear approximations you just might find
something, if that doesn't work try quadratic opprox.

**Comment by nobleliar9296** [comment]
also a suggestion try and implement what already in nature you'd be better
of.... also why are you fighting evolution? tell  me if you found this
useful.



-Enjoy

**Comment by TeHMoroS** [comment]
@nobleliar9296 I don't understand what's your point in those comments. I'm well aware of how derivatives and neural networks work and I'm not fighting anything. The derivatives are mathematically correct (as stated before), but the code logic of the network itself has an error, that's all (and I'm now 100% sure it is an error).

Let's look closer at the TanH activation function and it's derivative:

From neural2d-core.cpp:

    // tanh is a sigmoid curve scaled; output ranges from -1 to +1:
    float transferFunctionTanh(float x) { return tanh(x); }
    float transferFunctionDerivativeTanh(float x) { return 1.0f - tanh(x) * tanh(x); }

Those function are mathematically correct. Remember that those functions should be called with the "x" variable being exactly the same in each case. Now let's look at the feedForward() function:

    void Neuron::feedForward(Layer *pMyLayer)
    {
        float sum = 0.0;
        ... skipped summation code ...
        // Shape the output by passing it through the transfer function:
        float output = pMyLayer->tf(sum); // HERE - this is the place of interest
        this->output = output;
    }

As you see from above, the neuron output is the result of calling the transfer function. That is all correct so far. For example, when the weighted sum of neuron inputs equals 0.5, the neurons output with be the result of tanh(0.5) = 0.462 (rounded). Still correct.

Now on to backward propagation. Depending on the layer different functions are called, but they all call the TF derivative. Let's look at the hidden layer gradient descent's gradient calculation function:

    void Neuron::calcHiddenGradients(Layer &myLayer)
    {
        float dow = sumDOW_nextLayer(myLayer.pConnections);
        gradient = dow * myLayer.tfDerivative(output); // HERE - place of interest
    }

Fairly standard: get the weighted error from the next layer and multiply it by the output of the derivative function of TanH, the input being the current neuron output.

So, now what's happening here? Well, as mentioned earlier the function and it's derivative should always be called with the same input value, to give a mathematically correct, correlated pair of numbers, otherwise you calculate a one function (TF) for one value and another (TF derivative) for a different value. What I'm saying is that the TF derivative, as per the example above, should also be called with the input value being 0.5. Instead, it's called with the value of 0.462 - the output of the activation function. 0.5 and 0.462 are clearly different values, so you call a TF on one value and the derivative on another, which which makes the output not correlated, hence incorrect.

To put it in numbers:

    tanh(0.5) = 0.46211715726 // that's ok
    tanh'(0.5) = 0.78644773296 // that's what should be done, so the pair is correlated
    tanh'(0.46211715726) = 0.81354169517 // that's what Neural2D is doing - it basically generates a gradient for greater than in reality;
    // in relation to the original input (0.5) Neural2D calculates the derivative as such "return 1 - tanh(tanh(0.5)) * tanh(tanh(0.5));" - (the inner "tanh(0.5)" is the output of the neuron) which IS NOT the TanH's derivative

To cope with this, other neural network already take into account that the output neuron value has already a value AFTER calling TanH, so their derivative include that knowledge. The reimplementation of the TanH derivative that takes into account, that the passed input value is already including the output of TanH function itself should look like this:

    float transferFunctionDerivativeTanh(float x) { return 1.0f - x * x; } // x input is already a value of calling tanh(original_x)

If you still don't agree, then maybe the practical approach will convince you. The number of iterations needed to evaluate the NN for the XOR function is:
- original code: around 8700 iterations
- corrected code: 805 iterations

The numbers speak for themselves: 10 times faster network training to solve the problem, while maintaining all conditions (MSE included). Even my own experiment (160 inputs, 4 outputs, 5 layers, thousands of neurons) started behaving better after this fix.

**Comment by Codewander** [comment]
@TeHMoroS 
This is how I understand it:

Forward Propagation ( well understood :) )
1. This is known as Fan-In: from many(previous neurons) to one(current) 
2. Moving from left to right compute the output  for each neuron in the current Layer based on the outputs from previous layer through the associated weights
 output = Tanh(prodSum); 

BackPropagation:

void Neuron::calcHiddenGradients(Layer &myLayer)
{
    float dow = sumDOW_nextLayer(myLayer.pConnections);
    gradient = dow * myLayer.tfDerivative(output); // HERE - place of interest
}
0. calcHiddenGradients(Layer &myLayer) is called for each hidden layer.

1. We are moving from left to right (from output layer toward the input layer)
2. The output layer Gradient is computed uniquely for each neuron: Gradient = Error * TanhDeriv(Output);
3. Move left to the hidden layer. This becomes the current Layer(myLayer in the formal parameter code) with the current Neurons who's Gradients we want to compute!!!.
4. The next Layer is the layer to the right of the current layer. This layer has the Gradients already computed for each Neuron in the previous call.
5. WE WANT TO QUANTIFY, TO FIND OUT HOW MUCH DOES the current Neuron's DERIVATIVE OUTPUT (output computed during feed forward..) INFLUENCES the next Layer Neuron's Gradients through the respective weight links. That's why we take the Derivative of the Output of the current Neuron!!!!!!
6. This is also called Fan-Out: from one (current hidden Neuron) to many ( all the Neurons in next Layer connected to the current Neuron)
6. dow variable in the code is the product gradient sum (w * gradient) for each Neuron in the next Layer influenced by the current Neuron through the weight links.
7. SO FINALLY we multiply the product gradient sum with the Derivative of the current Neuron Output to get the Gradient of the current Neuron.
8.We do these steps for each Neuron in the current hidden Layer...

**Comment by TeHMoroS** [comment]
@Codewander Agreed on the actual logic of a NN. Still, in any mathematical model, when you calculate a value of a function for X and need a value of the function's derivative, you also calculate the derivative for X, not for Y. Neuron output is X before activation/transfer, but after that, it's not X anymore. It's Y - another value, for which you didn't call the original function (because it's the result of the said function), thus no mathematical correlation between them.

It's not a problem of how the NNs themselves are implemented, but a problem of basic correlation between two mathematical functions. TF(x) and TF'(x) are correlated, but TF(x) and TF'(y) are not. Plain and simple.
The derivative is used to measure the change to a function's value, given an input value of X. How do you expect it to measure efficiently, when you pass a different value to the derivative function (tanh(X) instead of X) than the one the original function was called with?

Theory aside, I repeat once again: after correcting the error the NN learns faster (10 times faster for the xor example bundled with the code, similar results for my own experiments). :)

**Comment by TeHMoroS** [comment]
@nobleliar9296 @Codewander Sorry, if I'm sounding abusive (or plainly rude). English is not my primary language and up to this point I haven't had a chance to use so much math-related terminology, which is a bit frustrating. ;)

I do have a proposal: you don't both of you (if you use the tanh or logistic TFs) try implementing the changes in your networks and see the results. Does it train faster? It it as accurate is before? The more different results we get, the better to resolve the issue (either by accepting or droping it).

The changes needed should be done in neural2d-core.cpp file:

**TanH**
From:

    float transferFunctionDerivativeTanh(float x) { return 1.0f - tanh(x) * tanh(x); }

To:

    float transferFunctionDerivativeTanh(float x) { return 1.0f - x * x; }

**Logistic**
From:

    float transferFunctionDerivativeLogistic(float x) { return exp(-x) / pow((exp(-x) + 1.0f), 2.0f); }

To:

    float transferFunctionDerivativeLogistic(float x) { return x * (1 - x); }

to include the fact, that the respective function's results are already in the neuron's output.

**Comment by davidrmiller** [comment]
Hi @TeHMoroS, Thank you so much for reporting your findings. I always had difficulty visualizing the backpropagation math and I could easily have made a mistake in that part of the code. If I remember correctly, one of the sources of my inspiration was the notes at http://www.cs.bham.ac.uk/~jxb/NN/nn.html . Would you be able to comment on the similarities or differences between the math in that version and mine?

Also thanks @Codewander for the useful comments. Unfortunately I'm saturated with other projects at this time and cannot give this issue the attention it deserves. I hope other mathematically-inclined readers can offer independent analyses. As @TeHMoroS suggested, it would be useful to see the results of using the proposed modification with a variety of data sets. Thanks to the community for help with this.

**Comment by TeHMoroS** [comment]
@davidrmiller I went through the notes you mentioned and it confirms the issue. Looking at the code with the backward propagation already included and looking at the hidden layer calculations I can confirm, that when propagating backwards, he also already takes into account that the TF has already been called.

Here's his feed forward for the hidden layer's neuron:
```c
    Hidden[p][j] = 1.0/(1.0 + exp(-SumH[p][j])) ;
```
Bare calculation:

    neuron_output = 1 / (1 + exp(-x))

And here's the backward propagation:
```c
    DeltaH[j] = SumDOW[j] * Hidden[p][j] * (1.0 - Hidden[p][j]) ;
```
Bare calculation:

    weighted_error = weighted_output_differences * neuron_output * (1 - neuron_output)

He's using sigmoid/logistic function derivative and there's no call to the exp() function itself, which is an indicator, that he's using the fact, than the neuron output already passed the TF.

Other than that, most of the math in Neural2D looks nice. Few things I found are related to the TFs:
- identity and linear TFs are the same thing (at the very least they give the same results); Their derivative is a constant 1;
- TanH and Logistic/Sigmoid is what I described a few comments above (should be simplified);
- ramp seems good;
- Gaussian and Softplus require to create a formula reversing the activation function first and combining it with their respective derivative functions;

Well, here goes:

**Gaussian**

    f(x) = exp(-1 * x^2 / 2)    f'(x) = -1 * x * exp(-1 * x^2 / 2)

First, let's reverse the TF. In context of the NN, "y" is the neuron output (result of calling the Gaussian TF):

    y = exp(-1 * x^2 / 2)    // let's use natural logarithm to reverse (inverse) the exponential
    ln(y) = ln(exp(-1 * x^2 / 2))
    ln(y) = -1 * x^2 / 2    // side swap and multiply by -1
    x^2 / 2 = -ln(y)
    x^2 = -2 * ln(y)
    x = sqrt(-2 * ln(y))    // that's our neuron output before the Gaussian TF

Now having this, let's substitute the "x" in the derivative:

    f'(x) = -1 * sqrt(-2 * ln(y)) * exp(-1 * sqrt(-2 * ln(y))^2 / 2)    // the sqrt(...)^2 are inverse towards each other, so let's shorten it a bit
    f'(x) = -1 * sqrt(-2 * ln(y)) * exp(-1 * -2 * ln(y) / 2)    // let's count the constants
    f'(x) = -1 * sqrt(-2 * ln(y)) * exp(ln(y))    // exp() and ln() are inverse, so let's shorten further
    f'(x) = -1 * sqrt(-2 * ln(y)) * y
    f'(x) = -y * sqrt(-2 * ln(y))    // whew, could someone please check if I haven't done a stupid mistake? ;)

By code, the Gaussian derivative for Neural2D would be:

```c
float transferFunctionDerivativeGaussian(float x) { return -x * sqrt(-2 * ln(x)); }
```

**Softplus**

    f(x) = ln(1 + exp(x))    f'(x) = 1 / (1 + exp(-x))

Reversing the TF function ("y" is the neuron output after calling the TF):

    y = ln(1 + exp(x))    // let's use the exponential to reverse (inverse) the natural logarithm
    exp(y) = exp(ln(1 + exp(x)))
    exp(y) = 1 + exp(x)    // side swap and substract 1 from both sides
    exp(x) = exp(y) - 1    //use natural logarithm to reverse (inverse) the exponential
    ln(exp(x)) = ln(exp(y) - 1)
    x = ln(exp(y) - 1)

The derivative has a troublesome `exp(-x)` in it, but we can do something about it:

    f'(x) = 1 / (1 + exp(-x)) = 1 / (1 + 1 / exp(x))

Now having this, let's substitute the "x" in the derivative:

    f'(x) = 1 / (1 + (1 / exp(ln(exp(y) - 1)))    // this got really ugly, really fast, luckily the first exp(ln(...)) are inverse
    f'(x) = 1 / (1 + (1 / (exp(y) - 1)))    // I think I'm going to use this one, but we can go further
    f'(x) = (exp(y) - 1) / exp(y)    // and again I hope I didn't make a stupid mistake anywhere

Depending on which of the last 2 lines is faster in code, that one should be used. Assuming the first of the two is faster the Softplus derivative for Neural2D would be:
```c
float transferFunctionDerivativeReLU(float x) { return 1 / (1 + (1 / (exp(x) - 1))); }
```
or eventually (for the second one):
```c
float transferFunctionDerivativeReLU(float x) { return (exp(x) - 1) / exp(x); }
```

Argh... my brain cells... It's been some time since I've done things like that :face_with_head_bandage: That was quite an exercise. :smile:
